File: android/src/test/java/de/idrinth/habitevaluator/android/FontSizeHelperTest.java
package de.idrinth.habitevaluator.android;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.content.res.Resources;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

class FontSizeHelperTest {

    private Context baseContext;
    private SharedPreferences prefs;
    private Resources resources;
    private Context configContext;

    @BeforeEach
    void setUp() {
        baseContext = mock(Context.class);
        prefs = mock(SharedPreferences.class);
        resources = mock(Resources.class);
        configContext = mock(Context.class);

        when(baseContext.getSharedPreferences(SettingsActivity.PREFS_NAME, Context.MODE_PRIVATE))
                .thenReturn(prefs);
        when(baseContext.getResources()).thenReturn(resources);
        when(resources.getConfiguration()).thenReturn(new Configuration());
        when(baseContext.createConfigurationContext(any(Configuration.class)))
                .thenReturn(configContext);
    }

    @Test
    void testSystemDefaultReturnsOriginalContext() {
        when(prefs.getString(SettingsActivity.KEY_FONT_SIZE, SettingsActivity.FONT_SIZE_SYSTEM))
                .thenReturn(SettingsActivity.FONT_SIZE_SYSTEM);

        Context result = FontSizeHelper.applyFontScale(baseContext);

        assertSame(baseContext, result);
        verify(baseContext, never()).createConfigurationContext(any());
    }

    @Test
    void testNullSettingReturnsOriginalContext() {
        when(prefs.getString(SettingsActivity.KEY_FONT_SIZE, SettingsActivity.FONT_SIZE_SYSTEM))
                .thenReturn(null);

        Context result = FontSizeHelper.applyFontScale(baseContext);

        assertSame(baseContext, result);
        verify(baseContext, never()).createConfigurationContext(any());
    }

    @Test
    void testUnknownSettingReturnsOriginalContext() {
        when(prefs.getString(SettingsActivity.KEY_FONT_SIZE, SettingsActivity.FONT_SIZE_SYSTEM))
                .thenReturn("UNKNOWN_VALUE");

        Context result = FontSizeHelper.applyFontScale(baseContext);

        assertSame(baseContext, result);
        verify(baseContext, never()).createConfigurationContext(any());
    }

    @Test
    void testXsFontSizeCreatesConfigurationContext() {
        when(prefs.getString(SettingsActivity.KEY_FONT_SIZE, SettingsActivity.FONT_SIZE_SYSTEM))
                .thenReturn(SettingsActivity.FONT_SIZE_XS);

        Context result = FontSizeHelper.applyFontScale(baseContext);

        assertSame(configContext, result);
        verify(baseContext).createConfigurationContext(argThat(config -> config.fontScale == 0.8f));
    }

    @Test
    void testSmallFontSizeCreatesConfigurationContext() {
        when(prefs.getString(SettingsActivity.KEY_FONT_SIZE, SettingsActivity.FONT_SIZE_SYSTEM))
                .thenReturn(SettingsActivity.FONT_SIZE_SMALL);

        Context result = FontSizeHelper.applyFontScale(baseContext);

        assertSame(configContext, result);
        verify(baseContext).createConfigurationContext(argThat(config -> config.fontScale == 0.9f));
    }

    @Test
    void testNormalFontSizeCreatesConfigurationContext() {
        when(prefs.getString(SettingsActivity.KEY_FONT_SIZE, SettingsActivity.FONT_SIZE_SYSTEM))
                .thenReturn(SettingsActivity.FONT_SIZE_NORMAL);

        Context result = FontSizeHelper.applyFontScale(baseContext);

        assertSame(configContext, result);
        verify(baseContext).createConfigurationContext(argThat(config -> config.fontScale == 1.0f));
    }

    @Test
    void testLargeFontSizeCreatesConfigurationContext() {
        when(prefs.getString(SettingsActivity.KEY_FONT_SIZE, SettingsActivity.FONT_SIZE_SYSTEM))
                .thenReturn(SettingsActivity.FONT_SIZE_LARGE);

        Context result = FontSizeHelper.applyFontScale(baseContext);

        assertSame(configContext, result);
        verify(baseContext).createConfigurationContext(argThat(config -> config.fontScale == 1.2f));
    }

    @Test
    void testCustomScaleDoesNotModifyOriginalConfiguration() {
        Configuration originalConfig = new Configuration();
        float originalFontScale = originalConfig.fontScale;
        when(resources.getConfiguration()).thenReturn(originalConfig);
        when(prefs.getString(SettingsActivity.KEY_FONT_SIZE, SettingsActivity.FONT_SIZE_SYSTEM))
                .thenReturn(SettingsActivity.FONT_SIZE_LARGE);

        FontSizeHelper.applyFontScale(baseContext);

        assertEquals(originalFontScale, originalConfig.fontScale, 0.001f);
    }
}

File: android/src/test/kotlin/de/idrinth/habitevaluator/android/FontSizeHelperTest.kt
package de.idrinth.habitevaluator.android

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import android.content.res.Resources
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.*

class FontSizeHelperTest {

    private lateinit var baseContext: Context
    private lateinit var prefs: SharedPreferences
    private lateinit var resources: Resources
    private lateinit var configContext: Context

    @BeforeEach
    fun setUp() {
        baseContext = mock(Context::class.java)
        prefs = mock(SharedPreferences::class.java)
        resources = mock(Resources::class.java)
        configContext = mock(Context::class.java)

        `when`(baseContext.getSharedPreferences(SettingsConstants.PREFS_NAME, Context.MODE_PRIVATE))
            .thenReturn(prefs)
        `when`(baseContext.resources).thenReturn(resources)
        `when`(resources.configuration).thenReturn(Configuration())
        `when`(baseContext.createConfigurationContext(any(Configuration::class.java)))
            .thenReturn(configContext)
    }

    @Test
    fun testSystemDefaultReturnsOriginalContext() {
        `when`(prefs.getString(SettingsConstants.KEY_FONT_SIZE, SettingsConstants.FONT_SIZE_SYSTEM))
            .thenReturn(SettingsConstants.FONT_SIZE_SYSTEM)

        val result = FontSizeHelper.applyFontScale(baseContext)

        assertSame(baseContext, result)
        verify(baseContext, never()).createConfigurationContext(any())
    }

    @Test
    fun testNullSettingReturnsOriginalContext() {
        `when`(prefs.getString(SettingsConstants.KEY_FONT_SIZE, SettingsConstants.FONT_SIZE_SYSTEM))
            .thenReturn(null)

        val result = FontSizeHelper.applyFontScale(baseContext)

        assertSame(baseContext, result)
        verify(baseContext, never()).createConfigurationContext(any())
    }

    @Test
    fun testUnknownSettingReturnsOriginalContext() {
        `when`(prefs.getString(SettingsConstants.KEY_FONT_SIZE, SettingsConstants.FONT_SIZE_SYSTEM))
            .thenReturn("UNKNOWN_VALUE")

        val result = FontSizeHelper.applyFontScale(baseContext)

        assertSame(baseContext, result)
        verify(baseContext, never()).createConfigurationContext(any())
    }

    @Test
    fun testXsFontSizeCreatesConfigurationContext() {
        `when`(prefs.getString(SettingsConstants.KEY_FONT_SIZE, SettingsConstants.FONT_SIZE_SYSTEM))
            .thenReturn(SettingsConstants.FONT_SIZE_XS)

        val result = FontSizeHelper.applyFontScale(baseContext)

        assertSame(configContext, result)
        verify(baseContext).createConfigurationContext(argThat { config -> config.fontScale == 0.8f })
    }

    @Test
    fun testSmallFontSizeCreatesConfigurationContext() {
        `when`(prefs.getString(SettingsConstants.KEY_FONT_SIZE, SettingsConstants.FONT_SIZE_SYSTEM))
            .thenReturn(SettingsConstants.FONT_SIZE_SMALL)

        val result = FontSizeHelper.applyFontScale(baseContext)

        assertSame(configContext, result)
        verify(baseContext).createConfigurationContext(argThat { config -> config.fontScale == 0.9f })
    }

    @Test
    fun testNormalFontSizeCreatesConfigurationContext() {
        `when`(prefs.getString(SettingsConstants.KEY_FONT_SIZE, SettingsConstants.FONT_SIZE_SYSTEM))
            .thenReturn(SettingsConstants.FONT_SIZE_NORMAL)

        val result = FontSizeHelper.applyFontScale(baseContext)

        assertSame(configContext, result)
        verify(baseContext).createConfigurationContext(argThat { config -> config.fontScale == 1.0f })
    }

    @Test
    fun testLargeFontSizeCreatesConfigurationContext() {
        `when`(prefs.getString(SettingsConstants.KEY_FONT_SIZE, SettingsConstants.FONT_SIZE_SYSTEM))
            .thenReturn(SettingsConstants.FONT_SIZE_LARGE)

        val result = FontSizeHelper.applyFontScale(baseContext)

        assertSame(configContext, result)
        verify(baseContext).createConfigurationContext(argThat { config -> config.fontScale == 1.2f })
    }

    @Test
    fun testCustomScaleDoesNotModifyOriginalConfiguration() {
        val originalConfig = Configuration()
        val originalFontScale = originalConfig.fontScale
        `when`(resources.configuration).thenReturn(originalConfig)
        `when`(prefs.getString(SettingsConstants.KEY_FONT_SIZE, SettingsConstants.FONT_SIZE_SYSTEM))
            .thenReturn(SettingsConstants.FONT_SIZE_LARGE)

        FontSizeHelper.applyFontScale(baseContext)

        assertEquals(originalFontScale, originalConfig.fontScale, 0.001f)
    }
}
